### DISCUSSION TITLE
Properties at every node of the course

### DISCUSSION THREAD

**Discussion by vishaldeltech** [Sunbird-Ed/Community#408] [General Discussion]
Hi folks

We are planning to add properties at every node of the course, to control various aspects of the course. Here are few examples

- Certifiable 
- Batchable
- Sellable
- Trackable
- Gradable

Here is a simple wireframe for the same
![Collection Properties](https://user-images.githubusercontent.com/67492325/219601473-42429c66-dd8c-4f4e-a5a0-95d69a7a93cb.jpg)

Additional info

- Categories will control the defaults for each property (set by tenant admins)
- If a new property is added to the system, then all courses in the system get the property with its default value
- In case of adding collection B to collection A, there are 2 scenarios: soft clone, and deep clone. In the case of a soft clone, the properties of collection B cannot be modified through course A. In the case of a deep clone, any property can be changed

#### What this allows

- Bundle of courses, where there may be certificates for individual course, and for the overall course
- Grading at a granular level
- Disable progress tracking for any node in a collection

Requesting feedback from the community  on the above approach

**Comment by Krishnaj20** [comment]
@vishaldeltech - the approach looks good to me. Can you expand more about these properties? for ex., what does batchable mean?

@maheshkumargangula @pallakartheekreddy @vrayulu @reshmi-nair @rhwarrier - Please go thru this and share your feedback.

**Reply by vishaldeltech** [reply]
Hi @Krishnaj20 Here are the details about properties

- Certifiable: Can the learner earn the certificate for completing this node
- trackable: Should progress of learner be tracked for items under this node
- Sellable: Should this node be visible on the e-commerce site for purchase
- Batchable: Can a batch be created for this node
- Gradable: Can the learner earn a grade for this node. Example a PGP course with 8 semesters, where there is grade for each semester, and a grade for the entire PGP course

There will be rules set-up for the properties, for example we cannot have trackable nodes under a non-trackable node

**Comment by vrayulu** [comment]
There are multiple enhancements being proposed in this thread:
- Adding new attributes (certifiable, sellable, etc)
- Ability to control these attributes at unit level
- Adding child collections in two ways: deep clone & soft clone

Each one may have to be addressed separately so that they can be generalised and used in multiple scenarios.

However, here are some of my thoughts:
- Is there any behaviour for each of the new attributes that is going to be implemented? if yes, what would be those behaviours and which building block will be implementing them?
- There is already ability to have different attribute values at each unit level in "knowlg" building block. Are we planning to have implementation in "Lern" building block to use these attributes while tracking the progress?
- Deep clone is already available via "copy" operation. There is also a capability of relation metadata using which we can enable metadata overloading when an object is added to a collection - this can be used to enable flexibility of using the same object in multiple different ways in different contexts.
- Looks like changes will be needed across multiple building blocks to implement these features. Please share if there are any thoughts on what will be implemented in each building block.
